### PR TITLE
Memory leak fixes - stream and filter life cycles

### PR DESCRIPTION
**Pull request by kumavis**
filters:
- filters are tied to the life cycle of their provider-engine/json-rpc-engine
- previous provider-engine never died so it kept everything
- now we create a mini json-rpc-engine for each page that does the filters, then passed on to our main stack
- that mini json-rpc-engine dies when the page closes

streams:
- we now clean up streams after the page disconnects

### CODE REVIEW

**Comment by kumavis** [comment]
This will likely need https://github.com/MetaMask/metamask-extension/pull/2069 to be merged in before tests will pass (seems we need newer versions of chrome/firefox?)

**Comment by kumavis** [comment]
I think the browser support issue is from `eth-json-rpc-filters` es7

File: app/scripts/contentscript.js [hunk 1]
```diff
@@ -50,22 +51,42 @@ function setupStreams () {
...
-  mx.ignoreStream('provider')
-  mx.ignoreStream('publicConfig')
+  mux.ignoreStream('provider')
+  mux.ignoreStream('publicConfig')
+}
+
+function logStreamDisconnectWarning (remoteLabel, err) {
+  let warningMsg = `MetamaskContentscript - lost connection to ${remoteLabel}`
+  if (err) warningMsg += '\n' + err.stack
+  console.warn(warningMsg)
```
Comment (flyswatter) [inline app/scripts/contentscript.js]: Prefer `log.warn()` for `loglevel` support.

File: app/scripts/lib/inpage-provider.js [hunk 2]
```diff
@@ -1,73 +1,56 @@
...
   const self = this
 
   // setup connectionStream multiplexing
-  var multiStream = self.multiStream = ObjectMultiplex()
-  pipe(
+  const mux = self.mux = new ObjectMultiplex()
+  pump(
     connectionStream,
-    multiStream,
+    mux,
```
Comment (flyswatter) [inline app/scripts/lib/inpage-provider.js]: I like that you've made the naming more uniform across files.

File: app/scripts/metamask-controller.js [hunk 3]
```diff
@@ -375,26 +385,64 @@ module.exports = class MetamaskController extends EventEmitter {
...
   }
 
   setupPublicConfig (outStream) {
-    pipe(
+    pump(
       this.publicConfigStore,
-      outStream
+      outStream,
+      (err) => {
+        if (err) console.error(err)
```
Comment (flyswatter) [inline app/scripts/metamask-controller.js]: Prefer `log.error()` for `loglevel` support.

File: app/scripts/metamask-controller.js [hunk 4]
```diff
@@ -375,26 +385,64 @@ module.exports = class MetamaskController extends EventEmitter {
...
+
     // log rpc activity
-    function onResponse (err, request, response) {
-      if (err) return console.error(err)
-      if (response.error) {
-        console.error('Error in RPC response:\n', response)
+    function loggerMiddleware (req, res, next, end) {
+      next((cb) => {
+        if (res.error) {
+          console.error('Error in RPC response:\n', res)
```
Comment (flyswatter) [inline app/scripts/metamask-controller.js]: Prefer `log.error()`, although I wouldn't block over this, as it was already a `console.error`.

**Review by danfinlay — APPROVED** [review]
This is honestly a very hard PR to review. I caught some linty issues, but nothing worth blocking over. I pulled the branch down, it continues to work. It would be cool to have some explanation of the exact leaks it fixes (I know there are filter leaks at several points in the lifecycle, but at least explaining here on the PR which spot this cleans up would be helpful).

Anyways, it works, doesn't break anything, cleans some things up, claims to fix a leaky realm of code, so I'm in favor of it, you can make changes I requested or merge as you like.

**Comment by kumavis** [comment]
the global log is only available in the background, not `inpage` and `contentscript`

**Comment by kumavis** [comment]
- [ ] breakout `originMiddleware`, `loggerMiddleware`, `createProviderMiddleware` to reduce MetamaskController file size
- [ ] disconnect filter-blockTracker handler on stream teardown

**Comment by kumavis** [comment]
@flyswatter updated!

**Pull request merged**